networkmanager/networkmanager.go

```diff
@@ -0,0 +1,265 @@
+
+	if err := manager.storage.RemoveNetworkInstanceInfo(instanceIdent); err != nil {
+		log.Errorf("Can't remove network info: %v", err)
+	}
+}
+
+// GetInstances gets instances.
+func (manager *NetworkManager) GetInstances() []aostypes.InstanceIdent {
+	manager.Lock()
+	defer manager.Unlock()
+
+	var instances []aostypes.InstanceIdent
+
+	for _, instancesData := range manager.instancesData {
+		for instanceIdent := range instancesData {
+			instances = append(instances, instanceIdent)
+		}
+	}
+
+	return instances
+}
+
+// PrepareInstanceNetworkParameters prepares network parameters for instance.
+func (manager *NetworkManager) PrepareInstanceNetworkParameters(
+	instanceIdent aostypes.InstanceIdent, networkID string,
+) (aostypes.NetworkParameters, error) {
+	networkParameters, err := manager.tryGetNetParametersFromCache(networkID, instanceIdent)
+	if err == nil {
+		return networkParameters, nil
+	}
+
+	var (
+		ip     net.IP
+		subnet *net.IPNet
+	)
+
+	defer func() {
+		if err != nil {
+			manager.deleteNetworkParametersFromCache(networkID, instanceIdent, ip)
+		}
+	}()
+
+	subnet, ip, err = GetIPSubnet(networkID)
+	if err != nil {
+		return networkParameters, err
+	}
+
+	networkParameters.IP = ip.String()
+	networkParameters.Subnet = subnet.String()
+
+	if networkParameters.VlanID, err = GetVlanID(networkID); err != nil {
+		return networkParameters, err
+	}
+
+	if err := manager.storage.AddNetworkInstanceInfo(NetworkInfo{
+		InstanceIdent:     instanceIdent,
+		NetworkID:         networkID,
+		NetworkParameters: networkParameters,
+	}); err != nil {
+		return networkParameters, aoserrors.Wrap(err)
+	}
+
+	manager.addNetworkParametersToCache(networkID, instanceIdent, networkParameters)
+
+	return networkParameters, nil
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func (manager *NetworkManager) deleteNetworkParametersFromCache(
+	networkID string, instanceIdent aostypes.InstanceIdent, ip net.IP,
+) {
+	manager.Lock()
+	defer manager.Unlock()
+
+	delete(manager.instancesData[networkID], instanceIdent)
+
+	if len(manager.instancesData[networkID]) == 0 {
+		manager.ipamSubnet.releaseIPNetPool(networkID)
+
+		return
+	}
+
+	manager.ipamSubnet.releaseIPToSubnet(networkID, ip)
+}
+
+func (manager *NetworkManager) addNetworkParametersToCache(
+	networkID string, instanceIdent aostypes.InstanceIdent, networkParameters aostypes.NetworkParameters,
+) {
+	manager.Lock()
+	defer manager.Unlock()
+
+	if _, ok := manager.instancesData[networkID]; !ok {
+		manager.instancesData[networkID] = make(map[aostypes.InstanceIdent]aostypes.NetworkParameters)
+	}
+
+	manager.instancesData[networkID][instanceIdent] = networkParameters
+}
+
+func (manager *NetworkManager) tryGetNetParametersFromCache(
```

Comment:
It would be better to have same pattern as for `addNetworkParametersToCache`, `deleteNetworkParametersFromCache`: `getNetworkParametersToCache`.

Reply:
Done